text/0000-impl-specialization.md

```diff
@@ -0,0 +1,1652 @@
+Limitations. But the rule is far less intuitive. And it suffers from some poor
+mathematical properties.
+
+Consider this example:
+
+```rust
+trait Trait1 {}
+trait Trait2: Trait1 {}
+trait Trait3: Trait2 {}
+
+trait Foo<T, U, V, W> {}
+
+impl<T, V>    Foo<T, T, V, V> for () where T: Trait1 {}   // impl I
+impl<T, U, V> Foo<T, U, U, V> for () where U: Trait2 {}   // impl J
+impl<T, U, V> Foo<T, U, V, V> for () where V: Trait3 {}   // impl K
+```
+
+This example is carefully designed so that `K <alt J <alt I`, but also `I <alt
+K`. That is, the `<=alt` relation fails to be transitive. Transitivity is
+important for the overall coherence of specialization.
+
+It's possible to work around the transitivity problem by rejecting any
+particular program that fails to have a transitive specialization. But the whole
+thing feels a bit hokey, and makes it much harder to tell at a glance whether
+one impl specializes another.
+
+All that said, it's possible to transition from the proposed `<=` rule to
+something like `<=alt` backwards-compatibly, since the latter allows strictly
+more specializations (i.e. only changes things in cases that would have
+previously generated an error).
+
+### Explicit ordering
+
+Another, perhaps more palatable alternative would be to take the specialization
+rule proposed in this RFC, but have some other way of specifying precedence when
+that rule can't resolve it -- perhaps by explicit priority numbering. That kind
+of mechanism is usually noncompositional, but due to the orphan rule, it's a
+least a crate-local concern. Like the alternative rule above, it could be added
+backwards compatibly if needed, since it only enables new cases.
+
+### Singleton non-default wins
+
+@pnkfelix suggested the following rule, which allows overlap so long as there is
+a unique non-default item.
+
+> For any given type-based lookup, either:
+>
+>  0. There are no results (error)
+>
+>  1. There is only one lookup result, in which case we're done (regardless of
+>     whether it is tagged as default or not),
+>
+>  2. There is a non-empty set of results with defaults, where exactly one
+>     result is non-default -- and then that non-default result is the answer,
+>     *or*
+>
+>  3. There is a non-empty set of results with defaults, where 0 or >1 results
+>     are non-default (and that is an error).
+
+This rule is arguably simpler than the one proposed in this RFC, and can
+accommodate the examples we've presented throughout. It would also support some
+of the cases this RFC cannot, because the default/non-default distinction can be
+used to specify an ordering between impls when the subset ordering fails to do
+so. For that reason, it is not forward-compatible with the main proposal in this
+RFC.
+
+The downsides are:
+
+- Because actual dispatch occurs at monomorphization, errors are generated quite
+  late, and only at use sites, not impl sites. That moves traits much more in
+  the direction of C++ templates.
```

Comment:
Ah, moving towards C++ templates was not my intent when I proposed this rule.  (That is, my goal was to try to simplify the reasoning for the programmer, but I did not realize such reasoning would have to be performed based on the fully-monomorphized types.)

So, consider any implicit support I might have had for this rule retracted.